code/controllers/voting.dm

```diff
@@ -14,345 +14,386 @@ datum/controller/vote
+	if(!C)
+		return
+	var/admin = check_rights(R_ADMIN, 0, user = C.mob)
+	voting |= C
+
+	var/dat = {"<script>
+		function update_vote_div(new_content) {
+			var votediv = document.getElementById("vote_div");
+			if(votediv) {
+				votediv.innerHTML = new_content;
+			}
+		}
+		</script>"}
+	if(mode)
+		dat += "<div id='vote_div'>[vote_html(C)]</div><hr>"
+		if(admin)
+			dat += "(<a href='?src=\ref[src];vote=cancel'>Cancel Vote</a>) "
+	else
+		dat += "<div id='vote_div'><h2>Start a vote:</h2><hr><ul><li>"
+		//restart
+		if(admin || config.allow_vote_restart)
+			dat += "<a href='?src=\ref[src];vote=restart'>Restart</a>"
 		else
-			. += "<h2>Start a vote:</h2><hr><ul><li>"
-			//restart
-			if(admin || config.allow_vote_restart)
-				. += "<a href='?src=\ref[src];vote=restart'>Restart</a>"
-			else
-				. += "<font color='grey'>Restart (Disallowed)</font>"
-			. += "</li><li>"
-			if(admin || config.allow_vote_restart)
-				. += "<a href='?src=\ref[src];vote=crew_transfer'>Crew Transfer</a>"
-			else
-				. += "<font color='grey'>Crew Transfer (Disallowed)</font>"
+			dat += "<font color='grey'>Restart (Disallowed)</font>"
+		dat += "</li><li>"
+		if(admin || config.allow_vote_restart)
+			dat += "<a href='?src=\ref[src];vote=crew_transfer'>Crew Transfer</a>"
+		else
+			dat += "<font color='grey'>Crew Transfer (Disallowed)</font>"
+		if(admin)
+			dat += "\t(<a href='?src=\ref[src];vote=toggle_restart'>[config.allow_vote_restart?"Allowed":"Disallowed"]</a>)"
+		dat += "</li><li>"
+		//gamemode
+		if(admin || config.allow_vote_mode)
+			dat += "<a href='?src=\ref[src];vote=gamemode'>GameMode</a>"
+		else
+			dat += "<font color='grey'>GameMode (Disallowed)</font>"
+		if(admin)
+			dat += "\t(<a href='?src=\ref[src];vote=toggle_gamemode'>[config.allow_vote_mode?"Allowed":"Disallowed"]</a>)"
+
+		dat += "</li>"
+		//custom
+		if(admin)
+			dat += "<li><a href='?src=\ref[src];vote=custom'>Custom</a></li>"
+		dat += "</ul></div><hr>"
+	var/datum/browser/popup = new(C.mob, "vote", "Voting Panel", nref=src)
+	popup.set_content(dat)
+	popup.open()
+
+/datum/controller/vote/proc/update_panel(var/client/client)
+	for(var/client/C in (client ? list(client) : voting))
+		C << output(url_encode(vote_html(C)), "vote.browser:update_vote_div")
+
+/datum/controller/vote/proc/vote_html(var/client/C)
+	. = ""
+	if(question)
+		. += "<h2>Vote: '[question]'</h2>"
+	else
+		. += "<h2>Vote: [capitalize(mode)]</h2>"
+	. += "Time Left: [time_remaining] s<hr><ul>"
+	for(var/i = 1, i <= choices.len, i++)
+		var/votes = choices[choices[i]]
+		if(!votes)
+			votes = 0
+		if(current_votes[C.ckey] == i)
+			. += "<li><b><a href='?src=\ref[src];vote=[i]'>[choices[i]] ([votes] vote\s)</a></b></li>"
+		else
+			. += "<li><a href='?src=\ref[src];vote=[i]'>[choices[i]] ([votes] vote\s)</a></li>"
+
+	. += "</ul>"
+
+
+/datum/controller/vote/Topic(href,href_list[],hsrc)
+	if(!usr || !usr.client)	return	//not necessary but meh...just in-case somebody does something stupid

```

Comment:
Oh god. Please don't nitpick all the code I haven't actually touched. I made a separate commit just for the code I actually changed, and even fixed some of those.